### ISSUE TITLE
Inverted Normal Maps

### ISSUE THREAD

**Issue by ghost** [TheDuckCow/MCprep#122]
# The Issue
When using a resource pack that supports normal maps, the green value from the normal map is incorrectly inputted into the normal of the principled shader. Blender uses the OpenGL format for normal maps, while shaders use an inverted green normal format (see fig. 1). I do not know why shaders treat the normal maps this way since shaders use OpenGL to render the game.

![normal_maps](https://user-images.githubusercontent.com/55573462/81880812-3bfceb80-9543-11ea-8fbb-98097ce73f76.png)
_Fig. 1. The two formats of normal maps, used in Blender and shaders, respectively. The normal map is of a pyramid-like structure making a bump on a surface._

As of now, MCprep simply plugs the normal map Image Texture node into the Normal Map node (see fig. 2). Since the green channel is inverted, this will render Y facing normals in an incorrect direction.

![Screen Shot 2020-05-13 at 6 20 31 PM](https://user-images.githubusercontent.com/55573462/81882288-32758280-9547-11ea-83cf-30ef93d4289c.png)
_Fig. 2. MCprep will plug the output of the normal map Image Texture node into the Normal Map node, without applying fixes to work with Blender._

# Potential Fixes
## Require Specific Resource Packs
Probably not a good solution, as many resource packs will have to change their normal mapping.
## Invert the Green Channel
This solution allows existing resource packs to work with Blender. To invert the green channel, the normal map's RGB values can be separated, then the green value can be inverted, then the values would be combined again. I have created this node setup for use with the SEUS PBR format (see fig. 3).

![Screen Shot 2020-05-13 at 6 20 55 PM](https://user-images.githubusercontent.com/55573462/81882450-a31c9f00-9547-11ea-9c0f-deb3f48b1c20.png)
_Fig. 3. A node setup in the SEUS PBR format with correct normal mapping_

## Other Notes
Another issue I have noticed is that imported normal textures have a color space of sRGB. MCprep in previous versions used to change this to a Non-Color color space, but this value was moved to the Image user, instead of on the node itself. This can cause issues with normal maps appearing black in some areas. I saw this in a previous issue, but it appears to have been closed and the issue still not resolved, so I am bringing this up again.

## Conclusions
The green value of the normal map needs to be inverted to be displayed correctly, and doing this in the node setup seems to be the best solution.

## Contributing
I am not fluent in Python, but I took a quick glance at the code, and it seems that the functions `matgen_cycles_principled`, `matgen_cycles_original`, and `matgen_cycles_emit` in `/MCprep/blob/master/MCprep_addon/materials/generate.py` should be edited to include the new node setup. 

3 new nodes will probably have to be defined, a separate RGB, combine RGB, and another invert node.
```python
# I am simply guessing what the nodes are called in the `nodes.new` function, I do not know the actual names.
nodeNormInvert = nodes.new('ShaderNodeInvert')
nodeNormSeparate = nodes.new('ShaderNodeSeparateRGB')
nodeNormCombine = nodes.new('ShaderNodeInvertCombineRGB')
```
Then these nodes would have to have their locations set. I am lazy and did not figure this out xD

Then these nodes would need to be connected.
```python
# I have not checked any of this, it is just my best guess and quick thoughts.
links.new(nodeTexNorm.outputs["Color"], nodeNormSeparate.inputs[0])
links.new(nodeNormSeparate.outputs["R"], nodeNormCombine.inputs[0])
links.new(nodeNormSeparate.outputs["G"], nodeNormInvert.inputs[1])
links.new(nodeNormSeparate.outputs["B"], nodeNormCombine.inputs[2])
links.new(nodeNormInvert.outputs["Color"], nodeNormCombine.inputs[1])
# Setting the invert value, I have no idea if the default value in Blender will apply.
nodeNormInvert.inputs[0]default_value = 1
```
Hopefully that helps! If you are busy during these times, let me know and I'll try my best to get a working fix in a PR. Also, my emails are a bit whack right now, so it may take me a 1-4 days to see a response.

**Comment by TheDuckCow** [comment]
Hi there! Mostly wanted to confirm receipt, and call out that I like the ideas.

Firstly, thank you so much for the detailed investigation and going as far as looking into the source files and mocking some code, very very helpful. And by all means, if you want to try working on a PR go for it - happy to offer advice or tips, especially if you're new to working with python.

Do you recall which github issue on color space it was? I'd like to re-open that issue if so. It should be [implemented here](https://github.com/TheDuckCow/MCprep/blob/master/MCprep_addon/materials/generate.py#L1016), but I confirm it's not working right in blender 2.82 now. Good catch.

I do have some time, but PBR is not my P0 item. Everything that you have said here and on the other issue https://github.com/TheDuckCow/MCprep/issues/78 (thanks for commenting there too) reinforce the necessity, at a minimum, of creating the "Sync materials operator". A lot of work will go into making pbr work right for most cases, but there will still be edge cases. My best case scenario is at least giving someone the option of only having to manually fix materials _once_ if the default behavior doesn't work. I realize I didn't log this enhancement issue before, [just did so now](https://github.com/TheDuckCow/MCprep/issues/124). Would love to hear any of your thoughts on that improvement as well, and the most useful way you think it could work.

**Comment by TheDuckCow** [comment]
Minor follow up, now I know what the issue is regarding applying non-color:

The texture swap operator, by design, does not completely regenerate materials. This is why currently prep materials creates the empty images slots for materials that aren't even using pbr passes  (muting as necessary).

But, with the new image-user setting, I'm not longer able to pre-assign what the image color space should be. You'll note that if you 1) swap texture pack and then 2) press prep materials, it then fixes the issue (has it will have regenerated the shader tree).

So the solution would be to make swap texturepacks actively search for and update the image user color space setting.

**Comment by devisionbit** [comment]
Thanks for your response!

The Issue on the change in the color space value was mentioned in issue #76 

I have tried your suggestion to first swap materials, then prep materials, however I noticed that on materials with only normal maps, the color space attribute stayed as `sRGB` while textures with spec maps were correctly done. Through some further looking, this is actually due to an if/else statement. 

See [line 1015-1018](https://github.com/TheDuckCow/MCprep/blob/a48b1b4c3a6660a6e5f392a0aed20454354866b8/MCprep_addon/materials/generate.py#L1015) of `generate.py`.
```python
if hasattr(nodeTexSpec, "color_space"): # 2.7 and earlier 2.8 versions
    nodeTexSpec.color_space = 'NONE'  # for better interpretation of specmaps
    nodeTexNorm.color_space = 'NONE'  # for better interpretation of normals
elif nodeTexSpec.image and hasattr(nodeTexSpec.image, "colorspace_settings"): # later 2.8 versions
    nodeTexSpec.image.colorspace_settings.name = 'Non-Color'
    if nodeTexNorm.image:
        nodeTexNorm.image.colorspace_settings.name = 'Non-Color'
```
This means only images with spec maps will have the ability to check and change the color space of normal maps. Instead, there should be separate if statements for each.
```python
if hasattr(nodeTexSpec, "color_space"): # 2.7 and earlier 2.8 versions
    nodeTexSpec.color_space = 'NONE'  # for better interpretation of specmaps
    nodeTexNorm.color_space = 'NONE'  # for better interpretation of normals
elif (nodeTexSpec.image and hasattr(nodeTexSpec.image, "colorspace_settings")) or (nodeTexNorm.image and hasattr(nodeTexNorm.image, "colorspace_settings")): # later 2.8 versions
    if nodeTexSpec.image and hasattr(nodeTexSpec.image, "colorspace_settings"):
        nodeTexSpec.image.colorspace_settings.name = 'Non-Color'
    if nodeTexNorm.image and hasattr(nodeTexNorm.image, "colorspace_settings"):
        nodeTexNorm.image.colorspace_settings.name = 'Non-Color'
```
I do now know exactly how Python handles multiple conditionals, and I am not exactly sure how to compile Blender add-ons, so I am not submitting a PR just yet, however, the code above should work according to my tiny JavaScript brain.

Regarding the Material Sync Operator, I am still trying to understand the overall functionality of it, so I will respond back once I do.

**Comment by TheDuckCow** [comment]
Good catch on the logic issue, I hadn't really thought of the case where only a normal *or* spec might be present, and not both at once.

I've updated this in https://github.com/TheDuckCow/MCprep/commit/33c4a8492e53208f41c0e1ffa5861ea2ffbcfec4 (planning to push this as a new release tonight... or tomorrow). 

This change above does NOT apply the logic around normal inversion, which I think is still valid, so feel free to take a stab at that.

I think it's worth exploring which node layout is better and more efficient in rendering (alt option below). I believe this setup below is equivalent and takes one less node, but as always it might come down to edge cases or cycles/eevee quirks. Worth doing some render tests, testing noise create and times, but largely should be a pretty fast implementation.

![invert green](https://user-images.githubusercontent.com/2958461/82003958-b440db00-962f-11ea-9b4d-160f29476ee8.png)

**Comment by devisionbit** [comment]
![screenshot](https://user-images.githubusercontent.com/55573462/82012263-3e874000-962c-11ea-874f-35e6f3a18680.png)

This was the first node setup that I tried, however, I was unsure about how this would translate to code through the API since Blender is not giving me any tooltips on the graph.

In terms of rendering times, they all scored within an error of margin (0.2 seconds). These were all in Cycles, as Eevee on my Mac seems to crash with many normal maps. I will try to do some testing tomorrow or the day after that on different rendering engines with a Windows 10 based PC. However, I believe the node setup above would work the best since it is cleaner and it only requires one node, so a slightly smaller file size.

**Comment by TheDuckCow** [comment]
Good call on that other approach. Don't fret about the testing, I'm happy to take the ownership of that (and functional is more important now that most optimal!).

Regarding this example RGB curve node, I did a little exploration - if you're curious about getting more into blender python, hopefully the steps below make sense!

----

## Find out how to modify a node with code!

### 1) Create a material (in the UI), add only the one node you care about, delete the others. Change the settings that you want to find
### 2) Drag the material id onto a py console window
Drop n drop is super useful for the console! Just drag the material icon to the left of the mat name; maybe type `mat =` first so you can reference it

### 3) get the one node with `node = mat.node_tree.nodes[0]
### 4) See what you can do with the node! `dir(node)` or just type `node.` and use auto complete in the py console.

At step four, I can see `node.type` = `'CURVE_RGB'`, and there is a `node.mapping` property, so looking at that further I see there is a thing called curves. I see it's a list, so I see how many `len(node.mapping.curves)` (answer is: 4, so based on UI looks like they might be combined, r,g,b), so now I check curves[0] vs curves[1] vs curves[2] to see if my inversion of the green channel is index [2]. I end up seeing the below:

```
node.mapping.curves[0].points[0].location >> Vector((0.0, 0.0))
node.mapping.curves[0].points[1].location >> Vector((1.0, 1.0))

# oddly this is the one which would seem to correspond to the green, instead of the third in list
node.mapping.curves[1].points[0].location >> Vector((0.0, 1.0))
node.mapping.curves[1].points[1].location >> Vector((1.0, 0.0))

node.mapping.curves[2].points[0].location >> Vector((0.0, 0.0))
node.mapping.curves[2].points[1].location >> Vector((1.0, 1.0))

``

I gave it a test by then running `node.mapping.curves[1].points[0].location = Vector((0.5, 0.5))`, and indeed this modes the point only in the channel channel curve. So it seems the order of the curves the node uses is r,g,b,combined. But! I'd need/want to test this is consistent across blender versions to be safe, so I'm still toying with wondering if safer to use the vector math nodes.

**Comment by devisionbit** [comment]
Wow, thanks so much for the tips! I forgot Blender came in with a Python console and all the developer features! I'll be sure to tinker around will all of this and more about the structure of the addon, I previously thought I was going to have to set up a Python environment, but Blender seems to have most of the necessary features.

Thank you again for the response!

**Comment by TheDuckCow** [comment]
Minor follow up on this issue, can you share whether this invert was changed in your master code? Having trouble keeping up with all the good things you are adding, a good problem to have!

**Comment by bytzo** [comment]
This issue was resolved in my development branch, [`bytzo:dev`](https://github.com/bytzo/MCprep/tree/dev) alongside all of the changes with material generation for PBR as referenced above in commit 30787ec. I realize now that this wasn't a good idea– as the commit would have to wait until all of the other changes in #127, leading to a longer time unresolved. 

About #127– I'm extremely sorry for the slow progress. I've been putting off [adding the non-principled generator](https://user-images.githubusercontent.com/55573462/82109241-51ac1580-96e9-11ea-9b01-24e17785393a.png) for quite some time now, and still need to ensure things work across Blender versions. At the moment, [`bytzo:dev`](https://github.com/bytzo/MCprep/tree/dev) seems to be quite stable and functional in 2.82 and up (I haven't gone back to test yet) except for the addition of a non-principled shader. Additionally, I am sure many things could be optimized, and I still need to work on fixing inconsistencies caused by my linter. 

If I can find time, I will finish up and make #127 stable across all versions tomorrow, so we can close this issue– it's been well over a month. Sorry for taking so long!

**Comment by TheDuckCow** [comment]
Marking as closed as I have merged the code into dev, will be released as a part of v3.2.4
